File: python/data_structures.py
import sys
from typing import Union
import math


class CircularBuffer:
    def __init__(self, max_capacity):
        self._max_capacity = max_capacity
        self._head = 0
        self._tail = 0
        self._size = 0
        self._buffer = [None] * self._max_capacity

    def size(self):
        return self._size

    def full(self):
        return self.size() == self._max_capacity

    def push_back(self, value: Union[int, float]):
        if self.full():
            self._buffer[self._head] = None
            self._head = (self._head + 1) % self._max_capacity
        else:
            self._size += 1
        self._buffer[self._tail] = value
        self._tail = (self._tail + 1) % self._max_capacity

    def pop_back(self):
        value = self._buffer[self._head]
        self._buffer[self._head] = None
        self._head = (self._head + 1) % self._max_capacity
        self._size -= 1
        return value

    def pop_front(self):
        value = self._buffer[self._tail]
        self._buffer[self._tail] = None
        self._tail = (self._tail + 1) % self._max_capacity
        self._size -= 1
        return value


# # DP- Longest sub list
# array = [1,8,4,2,7,3,7,5,3,7,9,12]

# def sequence(arr, sol_arr):
# 	if len(arr) == 0:
# 		return sol_arr

# 	excl = sequence(arr[1:], sol_arr)
# 	incl = []
# 	if len(sol_arr)==0 or arr[0] > sol_arr[-1]:
# 		sol_arr = sol_arr + [arr[0]]
# 		incl = sequence(arr[1:], sol_arr)

# 	if len(excl) > len(incl):
# 		return excl
# 	else:
# 		return incl

# print( sequence(array, []) )

# import queue
# class Graph:
# 	def __init__(self):
# 		self.graph = dict()
# 		self.node_ids = set()
# 		self.is_visited = set()

# 	def insert_edge(self,parent,children):
# 		if parent not in self.node_ids:
# 			self.node_ids.add(parent)
# 			self.graph[parent] = []
# 		for child in children:
# 			self.graph[parent].append(child)

# 	def BFS(self, root):
# 		self.is_visited = set()
# 		q = queue.Queue()
# 		q.put(root)
# 		while not q.empty():
# 			parent = q.get()
# 			if parent in self.is_visited:
# 				continue
# 			self.is_visited.add(parent)
# 			if parent in self.node_ids:
# 				children = self.graph[parent]
# 				for child in children:
# 					q.put(child)
# 					print(child)

# 	def DFS_util(self,q):
# 		parent = q.pop()
# 		if parent in self.is_visited:
# 			return
# 		self.is_visited.add(parent)
# 		if parent in self.node_ids:
# 			children = self.graph[parent]
# 			for child in children:
# 				q.append(child)
# 				print(child)
# 				self.DFS_util(q)

# 	def DFS(self,root):
# 		q = []
# 		q.append(root)
# 		self.is_visited = set()
# 		self.DFS_util(q)


# g = Graph()
# g.insert_edge(0,[1,2,3])
# g.insert_edge(1,[4])
# g.insert_edge(2,[4,5,6])
# g.insert_edge(3,[5,6])
# g.insert_edge(5,[7,8])
# g.insert_edge(6,[8])

# # g.BFS(0)
# g.DFS(0)

# class node:
# 	def __init__(self,data):
# 		self.data = data
# 		self.next_node = None

# class ll:
# 	def __init__(self):
# 		self.head = None
# 		self.tail = None

# 	def insert_node(self,data):
# 		new_node = node(data)
# 		if self.head is None:
# 			self.head = new_node
# 			self.tail = new_node
# 		else:
# 			self.tail.next_node = new_node
# 			self.tail = new_node

# linked_list = ll()
# linked_list.insert_node(1)
# linked_list.insert_node(2)
# linked_list.insert_node(3)
# linked_list.insert_node(4)

# current_node = linked_list.head
# while current_node.next_node != None:
# 	print(current_node.data)
# 	current_node = current_node.next_node


# Simple class and object
# class student:
# 	def __init__(self,first_name,last_name):
# 		self.first_name = first_name
# 		self.last_name = last_name
# 		return

# 	def logger(self):
# 		print(self.first_name + self.last_name)

# # Main
# a = student("xyz","abc")
# a.logger()
# student.logger(a)


# LINKED lists
# Linked arrays or linked lists use pointer to next element and are not stored in contiguous locations in memory
# drawback is, to access an element, we have to go sequentially. A binary search cannot be conducted
# Not cache friendly

# # Creating a list of objects of specific class
# class node:
# 	child = None
# 	def __init__(self,data):
# 		self.x = data[0]
# 		self.y = data[1]
# 		self.z = data[2]
# 		return

# # Separate linked list class can also be used to create a head which is node object and then the same process.
# class linked_list:
# 	def __init__(self,parent):
# 		self.parent = parent


# data1 = [0,0,0]
# data2 = [0.5,0.5,0.5]
# data3 = [1,1,1]

# nodes1 = node(data1)
# nodes2 = node(data2)
# nodes3 = node(data3)

# nodes1.child = nodes2
# nodes2.child = nodes3

# print(nodes1.child)
# print(nodes2.child)
# print(nodes3.child)


# # Python LISTS
# # a = [1,2,3,4]
# # b = [[1,2,3],[4,5,6]] # Multi dimensional list

# # OR

# # Append(element)
# a = []
# a.append(1)
# a.append(2)
# a.append((3,4)) # This will append a tuple (3,4) to the list. Basically things get appended as it is
# a.append("string")
# print(a)
# print(a[2][0]) # tuple elements can be accessed like this.

# # Insert(pos,element)
# b = []
# b.insert(1,4)
# print(b[0])
# # print(b[1]) # This will print list index out of range

# b = [1,2,3,4]
# b.insert(1,5) # This will append element at 1 position so new list will be 1,5,2,3,4
# print(b)

# # Extend([elements]) method appends multiple elements at the end
# a = [1,2]
# a.extend([3,4,5])
# print(a)

# # List elements are accessed using [index]. A negative index can be used to access elements from the end.
# # list.remove(element) removes the corresponding element from the list
# a = [1,2,3,4]
# b = a.pop()
# print(a)
# print(b)

# a = [1,2,3,4]
# b = a.pop(2)
# print(a)
# print(b)

# # Slicing operation can be performed on the list like a[3:8] to print elements from index 3 to 7
# a = [1,2,3,4,5,6]
# print(a[2:5])

# # reverse, sort, count, and copy are other useful methods


# from collections import defaultdict
# import queue

# # Breadth first search
# class Graph:
# 	def __init__(self):
# 		self.graph = defaultdict(list)
# 		return

# 	def addEdge(self,parent,child):
# 		self.graph[parent].append(child)
# 		return

# def BFS(graph,strt_node):
# 	print("Running BFS")
# 	isVisited = [False] * len(graph.graph) # Creating a list of specific size
# 	q = queue.LifoQueue() # LIFO
# 	q.put(strt_node)
# 	isVisited[strt_node] = True
# 	while(not q.empty()):
# 		curr_node = q.get()
# 		print("Current Node: {}". format(curr_node))
# 		neighbrs = graph.graph[curr_node]
# 		for i in neighbrs:
# 			if not isVisited[i]:
# 				q.put(i)
# 				isVisited[i] = True

# 	return


# if __name__ == '__main__':
# 	g = Graph()
# g.addEdge(0, 1)
# g.addEdge(0, 2)
# g.addEdge(1, 2)
# g.addEdge(2, 0)
# g.addEdge(2, 3)
# g.addEdge(3, 3)
# 	BFS(g,2)

File: Cplusplus/src/data_structures.cpp
#include <iostream>
// #include <list>

// // Lists. Check methods in geeks for geeks
// void print_list(std::list <int> ip_list)
// {
// 	std::list <int> :: iterator it;
// 	for(it=ip_list.begin(); it!=ip_list.end();++it)
// 	{
// 		std::cout<< *it << "\t";
// 	}
// 	std::cout<< std::endl;
// }
// int main(int argc, char** argv)
// {
// 	/* Defined as list <data type> similar to vectors.
// 	Operations like push_back and push_front are less expensive compared to vectors
// 	*/
// 	std::list <int> a;
// 	a.push_back(0);
// 	a.push_back(1);
// 	a.push_back(2);	
// 	print_list(a);
// 	return 0;
// }








// // Linked lists
// // dot operator is used to access values of struct or members and -> is used to access them using pointers.
// struct node
// {
// 	int data;
// 	node *next;
// };

// class SinglyLinkedList
// {
// public:
// 	node* head;
// 	node* tail;
// public:
// 	SinglyLinkedList()
// 	{
// 		head = NULL;
// 		tail = NULL;
// 	};

// 	void addNode(int _data)
// 	{
// 		node* temp = new node; // If new node not present, memory is not allocated and it will throw an error when we try to store value for this new node.
// 		temp->data = _data;
// 		temp->next = NULL;
// 		if (!head)
// 		{
// 			// std::cout<< "head is null" << std::endl;
// 			head = temp;
// 			tail = temp;
// 		}
// 		else
// 		{
// 			// std::cout<< "head is not null" << std::endl;
// 			tail->next = temp;
// 			tail = temp;
// 		}
// 	};

// 	void printList()
// 	{
// 		node* temp = this->head;
// 		while(temp)
// 		{
// 			std::cout<< "Data is: " << temp->data << std::endl;
// 			temp = temp->next;
// 		}

// 	};
// };

// int main(int argc, char** argv)
// {
// 	SinglyLinkedList llist;
// 	for (size_t i=0; i<100; ++i)
// 		llist.addNode(i);
	
// 	llist.printList();
// 	return 0;
// }










// // Breadth first search and Depth First Search
// // API for graph
// #include <vector>
// #include <queue>
// #include <stack>
// class Graph
// {
// public:
// 	std::vector<std::vector<int>> graph;

// public:
// 	Graph(){}
// 	void addEdge(int parent, int child)
// 	{
// 		if (parent >= graph.size())
// 		{
// 			graph.resize(parent+1);
// 			graph[parent].push_back(child);
// 		}
// 		else
// 			graph[parent].push_back(child);
// 	};

// 	void printGraph()
// 	{
// 		for (int i=0; i < graph.size(); ++i)
// 		{
// 			std::cout<< i << ": ";
// 			for (int j=0; j < graph[i].size(); ++j)
// 			{
// 				if (j==graph[i].size()-1)
// 					std::cout<< graph[i][j] << std::endl;
// 				else
// 					std::cout<< graph[i][j] << " , ";
// 			}				
// 		}
// 	}
// };


// class Searches
// {
// private:
// public:
// 	Searches(){};

// 	void BFS(std::vector<std::vector<int>> graph, int strt_node)
// 	{
		// std::vector<bool> isVisited;
		// for (int i=0; i < graph.size(); ++i)
		// {
		// 	for (int j=0; j < graph[i].size(); ++j)
		// 	{
		// 		if (graph[i][j]>=isVisited.size())
		// 		{
		// 			isVisited.resize(graph[i][j]+1);
		// 			isVisited[graph[i][j]] = false;
		// 		}
		// 		else
		// 			isVisited[graph[i][j]] = false;
		// 	}
		// }

// 		std::queue<int> q;
// 		q.push(strt_node);
// 		isVisited[strt_node] = true;
// 		while(!q.empty())
// 		{
// 			int curr_node = q.front();
// 			q.pop();
// 			std::cout<< curr_node << std::endl;
// 			std::vector<int> neighbrs = graph[curr_node];
// 			for (int i=0; i<neighbrs.size();++i)
// 			{
// 				if (!isVisited[neighbrs[i]])
// 				{
// 					q.push(neighbrs[i]);
// 					isVisited[neighbrs[i]] = true;
// 				}
// 			}
// 		}
// 	};

// 	void DFS_util(std::vector<std::vector<int>> graph,std::vector<bool> isVisited,std::stack<int> s)
// 	{
// 		int curr_node = s.top();
// 		s.pop();
// 		std::cout<< curr_node << std::endl;
// 		std::vector<int> neighbrs = graph[curr_node];
// 		for (int i=0; i<neighbrs.size();++i)
// 		{
// 			if (!isVisited[neighbrs[i]])
// 			{
// 				s.push(neighbrs[i]);
// 				isVisited[neighbrs[i]] = true;
// 				DFS_util(graph,isVisited,s);	
// 			}
// 		}

// 	};
// 	void DFS(std::vector<std::vector<int>> graph,int strt_node)
// 	{
// 		std::vector<bool> isVisited;
// 		for (int i=0; i < graph.size(); ++i)
// 		{
// 			for (int j=0; j < graph[i].size(); ++j)
// 			{
// 				if (graph[i][j]>=isVisited.size())
// 				{
// 					isVisited.resize(graph[i][j]+1);
// 					isVisited[graph[i][j]] = false;
// 				}
// 				else
// 					isVisited[graph[i][j]] = false;
// 			}
// 		}
// 		std::stack<int> s;
// 		s.push(strt_node);
// 		isVisited[strt_node] = true;
// 		DFS_util(graph,isVisited,s);	
// 	};
// };


// int main(int argc, char** argv)
// {
// 	Graph g;
// 	g.addEdge(0, 1);
// 	g.addEdge(0, 2);
// 	g.addEdge(1, 2); 
// 	g.addEdge(2, 0); 
// 	g.addEdge(2, 3); 
// 	g.addEdge(3, 3);
// 	// g.printGraph();

// 	Searches traverser;

// 	// Run BFS
// 	traverser.BFS(g.graph,2);
// 	traverser.DFS(g.graph,2);
// 	return 0;
// }










// Binary trees
#include <queue>
struct binary_tree
{
	int data;
	binary_tree* left;
	binary_tree* right;
	binary_tree* parent;
};

class tree
{
private:
	binary_tree* root_node;
	std::vector<bool> isVisited;
	int no_nodes;
public:
	tree() {root_node = NULL;};

	binary_tree* create_node(int _data)
	{
		binary_tree* node = new binary_tree;
		node->data = _data;
		node->parent = NULL;
		node->left = NULL;
		node->right = NULL;
		return node;
	};
	void addConnection(binary_tree* node, binary_tree* parent, binary_tree* left_child, binary_tree* right_child)
	{
		if (!root_node)
			root_node = node;
		node->left = left_child;
		node->right = right_child;
		node->parent = parent;

	};
	void init_visited()
	{
		no_nodes = 0;		
		std::queue<binary_tree*> q;
		q.push(root_node);
		no_nodes+=1;
		while(!q.empty())
		{
			binary_tree* temp = q.front(); q.pop();
			if (temp->left)
			{
				q.push(temp->left);
				no_nodes+=1;
			}
		
			if (temp->right)
			{
				q.push(temp->right);
				no_nodes+=1;
			}
			
		}
		std::cout<< "Number of nodes: " << no_nodes << std::endl;
		isVisited.resize(no_nodes);
		for (int i=0; i < no_nodes; ++i)
			isVisited[i] = false;
	};
	void printLevelOrder()
	{
		std::queue<binary_tree*> q;

		q.push(root_node);
		while(!q.empty())
		{
			binary_tree* temp = q.front(); q.pop();
			if (!temp->left && !temp->right)
			{
				std::cout<< "Leaf Node: " << temp->data << std::endl;
				continue;
			}

			std::cout<< "Parent: " << temp->data << " Children: ";
			if (temp->left)
			{
				std::cout<< "Left: " << temp->left->data << " ";
				q.push(temp->left);
			}
		
			if (temp->right)
			{
				std::cout<< "Right: " << temp->right->data << " ";
				q.push(temp->right);
			}
			std::cout<< std::endl;
		};
	};


	void preorder_util(binary_tree* node)
	{
		if (!isVisited[node->data-1])
		{
			// print the current node
			std::cout<< node->data << std::endl;
			isVisited[node->data-1] = true;
		}

		if (node->left!=NULL)
			preorder_util(node->left);
		if (node->right!=NULL)
			preorder_util(node->right);
	}
	void preorder() // nlr version of DFS
	{
		init_visited();
		preorder_util(root_node);	
	};

	void inorder_util(binary_tree *node) //lnr
	{
		if (node->left!=NULL)
			inorder_util(node->left);
		if (!isVisited[node->data-1])
		{
			std::cout<< node->data << std::endl;
			isVisited[node->data-1] = true;
		}
		if (node->right!=NULL)
			inorder_util(node->right);
	};
	void inorder()
	{
		init_visited();
		inorder_util(root_node);
	};

	void postorder_util(binary_tree* node)
	{
		if (node->left!=NULL)
			postorder_util(node->left);
		if (node->right!=NULL)
			postorder_util(node->right);
		if (!isVisited[node->data-1])
		{
			std::cout<< node->data << std::endl;
			isVisited[node->data-1] = true;
		}	
	};
	void postorder()
	{
		init_visited();
		postorder_util(root_node);
	};
};


int main(int argc, char** argv)
{

	// Option-1
	// binary_tree node1;
	// binary_tree node2;
	// binary_tree node3;
	// node1.data = 1;
	// node2.data = 2;
	// node3.data = 3;

	// node1.left = &node2;
	// node1.right = &node3;
	// node1.parent = &node1;

	// std::cout<< node1.left->data << std::endl;
	// std::cout<< node1.right->data << std::endl;
	// std::cout<< node1.parent->data << std::endl;


	// Option-2. Creating memory in the heap. We can keep creating pointers to different nodes and then write a function to add edge
	// which basically adds connection pointing the parent to the child.
	tree bin_tree;
	binary_tree* node1 = bin_tree.create_node(1);
	binary_tree* node2 = bin_tree.create_node(2);
	binary_tree* node3 = bin_tree.create_node(3);
	binary_tree* node4 = bin_tree.create_node(4);
	binary_tree* node5 = bin_tree.create_node(5);
	// binary_tree* node6 = bin_tree.create_node(6);
	
	// Some Tree-1
	// bin_tree.addConnection(node1, NULL, node2, node3);
	// bin_tree.addConnection(node2, node1, node4, NULL);
	// bin_tree.addConnection(node3, node1, node5, node6);


	// Tree 2 and some traversals
	bin_tree.addConnection(node1,NULL,node2,node3);
	bin_tree.addConnection(node2,node1,node4,node5);
	bin_tree.printLevelOrder();
	std::cout<< "Preorder" << std::endl;
	bin_tree.preorder();
	std::cout<< std::endl<< std::endl;
	std::cout<< "Inorder" << std::endl;
	bin_tree.inorder();
	std::cout<< std::endl<< std::endl;
	std::cout<< "Postorder" << std::endl;
	bin_tree.postorder();


	// Tree -3
	
	return 0;
}




